python/ray/_private/node.py

```diff
@@ -964,6 +972,120 @@ def _get_cached_port(
+        """
+        for i in range(timeout_s):
+            node_ip_address = self._get_cached_node_ip_address()
+
+            if node_ip_address is not None:
+                return node_ip_address
+
+            time.sleep(1)
+            if i % 10 == 0:
+                logger.info(
+                    "Can't find a `node_ip_address.json` file from "
+                    f"{self.get_session_dir_path()}. "
+                    "Have you started Ray instsance using "
+                    "`ray start` or `ray.init`?"
+                )
+
+        raise ValueError(
+            "Can't find a `node_ip_address.json` file from "
+            f"{self.get_session_dir_path()}. "
+            f"for {timeout_s} seconds. "
+            "A ray instance hasn't started. "
+            "Did you do `ray start` or `ray.init` on this host?"
+        )
+
+    def _get_cached_node_ip_address(self) -> str:
+        """Get a node address cached on this session.
+
+        If a ray instance is started by `ray start --node-ip-address`,
+        the node ip address is cached to a file node_ip_address.json.
+        Otherwise, the file exists, but it is emptyl.
+
+        This API is process-safe, meaning the file access is protected by
+        a file lock.
+
+        Returns:
+            node_ip_address cached on the current node. None if the node
+            the file doesn't exist, meaning ray instance hasn't been
+            started on a current node. If node_ip_address is not written
+            to a file, it means --node-ip-address is not given, and in this
+            case, we find the IP address ourselves.
+        """
+        assert hasattr(self, "_session_dir")
+        file_path = Path(
+            os.path.join(self.get_session_dir_path(), "node_ip_address.json")
+        )
+        cached_node_ip_address = {}
+
+        with FileLock(str(file_path.absolute()) + ".lock"):
+            if not file_path.exists():
+                return None
+
+            with file_path.open() as f:
+                cached_node_ip_address.update(json.load(f))
+
+            if "node_ip_address" in cached_node_ip_address:
+                return cached_node_ip_address["node_ip_address"]
+            else:
+                return ray.util.get_node_ip_address()
+
+    def _write_node_ip_address(self, node_ip_address: Optional[str]) -> None:
+        """Write a node ip address of the current session to
+        node_ip_address.json.
+
+        If a ray instance is started by `ray start --node-ip-address`,
+        the node ip address is cached to a file node_ip_address.json.
+
+        This API is process-safe, meaning the file access is protected by
+        a file lock.
+
+        The file contains a single string node_ip_address. It nothing
+        is written, --node-ip-address is not given. In this case, Ray
+        resolves the IP address on its own. It assumes in a single node,
+        you can have only 1 IP address (which is the assumption ray
```

Comment:
This might not be true in the future. the gRPC can be binded to multiple address.
but it's true for now.

Reply:
I think when we support multiple ip, this will be broken, and we can fix it at that time.